src/Entity/Citation.php

```diff
@@ -0,0 +1,377 @@
+      // This will pull out the `<a href....>` part of the link.
+      preg_match('/<a.*?>/', (string) $link->toString(), $matches);
+      return $matches[0] ?? NULL;
+    }
+  }
+
+  /**
+   * Get the label of the entity wrapped in a link tag to the parent or url.
+   *
+   * @return \Drupal\Core\Link|null
+   *   Label or linked label string.
+   *
+   * @throws \Drupal\Core\Entity\EntityMalformedException
+   */
+  protected function getLink() {
+    $url = NULL;
+
+    // Link to the parent node.
+    if ($parent_entity = $this->getParentEntity()) {
+      $url = $parent_entity->toUrl();
+    }
+
+    // The user entered url.
+    if ($url_string = $this->getUrl()) {
+      $url = $this->getUrlFromString($url_string) ?? $url;
+    }
+
+    // The doi field is a string, so construct the full url from that.
+    if ($doi = $this->getDoi()) {
+      $url = $this->getUrlFromString("https://doi.org/$doi") ?? $url;
+    }
+
+    if ($url) {
+      return Link::fromTextAndUrl('[replace]', $url);
+    }
+  }
+
+  /**
+   * Get a url object from the provided url/string.
+   *
+   * @param string $string
+   *   User entered string.
+   *
+   * @return \Drupal\Core\Url|null
+   *   Url object if successful.
+   */
+  protected function getUrlFromString($string): ?Url {
+    try {
+      return Url::fromUserInput($string);
+    }
+    catch (\Exception $e) {
+      try {
+        return Url::fromUri($string);
+      }
+      catch (\Exception $e) {
+        // Nothing to do, just fall back to returning null.
+      }
+    }
+    return NULL;
+  }
+
+  /**
+   * Fallback function to get the entity field's string value.
+   *
+   * @param string $name
+   *   Function name.
+   * @param mixed $args
+   *   Args.
+   *
+   * @return string
+   *   Entity field value as a string.
+   */
+  public function __call($name, $args) {
+    // Remove the `get` from the beginning.
+    $data_name = preg_replace('/^get/', '', $name);
+
+    // Convert UpperCamelCase to snake_case. This allows us to dynamically
+    // fetch field names just by using the method names. Later versions it would
+    // be preferred to have a UI that allows the user to choose which field
+    // maps to which variable in the CSL.
+    preg_match_all('!([A-Z][A-Z0-9]*(?=$|[A-Z][a-z0-9])|[A-Za-z][a-z0-9]+)!', $data_name, $matches);
+    $ret = $matches[0];
+    foreach ($ret as &$match) {
+      $match = $match == strtoupper($match) ? strtolower($match) : lcfirst($match);
+    }
+    $data_name = implode('_', $ret);
+
+    if ($field = $this->getFieldName($data_name)) {
+      return $this->get($field)->getString();
+    }
+  }
+
+  /**
+   * Get the list of authors from the entity field.
+   *
+   * @return array|null
+   *   Keyed array of author data.
+   */
+  protected function getAuthor() {
+    // Authors are rendered using `names` render element in CSL. It expects
+    // the name array to be keyed with `given`, `family` etc keys. Luckily the
+    // name module does this for us.
```

Comment:
This is cool.  Names are one of those things that seems simple, and ends up being super hard.  The names module is a good approach.  Do we use it in the `stanford_person` module as well?

Reply:
we don't at the moment. but that would be a really nice enhancement I think.